code/modules/antagonists/mind_flayer/flayer_datum.dm

```diff
@@ -0,0 +1,319 @@
+	var/old_swarm_amount = usable_swarms
+	usable_swarms = amount
+	if(update_total)
+		var/difference = usable_swarms - old_swarm_amount
+		if(difference < 0)
+			// Otherwise buying an ability can reduce your `total_swarms_gathered`
+			return
+		total_swarms_gathered += difference
+
+/*
+* amount - The positive or negative number to adjust the swarm count by, result clamped above 0
+*/
+/datum/antagonist/mindflayer/proc/adjust_swarms(amount, bound_lower = 0, bound_upper = INFINITY)
+	set_swarms(clamp((usable_swarms + amount), bound_lower, bound_upper), TRUE)
+
+//This is mostly for flavor, for framing messages as coming from the swarm itself. The other reason is so I can type "span" less.
+/datum/antagonist/mindflayer/proc/send_swarm_message(message)
+	if(HAS_TRAIT(owner.current, TRAIT_MINDFLAYER_NULLIFIED))
+		message = stutter(message, 0, TRUE)
+	to_chat(owner.current, "<span class='sinister'>[message]</span>")
+
+/**
+	Checks for any reason that you should not be able to drain someone for.
+	Returns either true or false, if the harvest will work or not.
+*/
+/datum/antagonist/mindflayer/proc/check_valid_harvest(mob/living/carbon/human/H)
+	if(HAS_TRAIT(owner.current, TRAIT_MINDFLAYER_NULLIFIED))
+		send_swarm_message("We do not have the energy for this...")
+		return FALSE
+	if(IS_MINDFLAYER(H))
+		send_swarm_message("Their hive rejects the connection.")
+		return FALSE
+	var/obj/item/organ/internal/brain/brain = H.get_int_organ(/obj/item/organ/internal/brain)
+	if(!istype(brain))
+		send_swarm_message("This entity has no brain to harvest from.")
+		return FALSE
+	if(!H.ckey && !H.player_ghosted)
+		send_swarm_message("This brain does not contain the spark that feeds us. Find more suitable prey.")
+		return FALSE
+	if(brain.damage >= brain.max_damage || (H.stat == DEAD && !H.has_status_effect(STATUS_EFFECT_RECENTLY_SUCCUMBED)))
+		send_swarm_message("We detect no neural activity to harvest from this brain.")
+		return FALSE
+	if(HAS_MIND_TRAIT(H, TRAIT_XENOBIO_SPAWNED_HUMAN))
+		send_swarm_message("This brain is unsuitable to harvest.")
+		return FALSE
+
+	var/unique_drain_id = H.UID()
+	if(isnull(drained_humans[unique_drain_id]))
+		drained_humans[unique_drain_id] = 0
+	else if(drained_humans[unique_drain_id] > BRAIN_DRAIN_LIMIT)
+		if(!has_notified)
+			has_notified = TRUE
+			send_swarm_message("You have drained most of the life force from [H]'s brain, and you will get no more swarms from them!")
+		return DRAIN_BUT_NO_SWARMS
+	return TRUE
+
+/**
+	Begins draining the brain of H, gains swarms equal to the amount of brain damage dealt per tick. Upgrades can increase the amount of damage per tick.
+**/
+/datum/antagonist/mindflayer/proc/handle_harvest(mob/living/carbon/human/H)
+	harvesting = H
+	var/obj/item/organ/internal/brain/drained_brain = H.get_int_organ(/obj/item/organ/internal/brain)
+	var/unique_drain_id = H.UID()
```

Comment:
I think you need a null/`istype` check on `drained_brain` here, otherwise you'll runtime if they're brainless and you hit the visible message

Reply:
We already check that in `check_valid_harvest()`!